homeassistant/components/duwi/__init__.py

```diff
@@ -0,0 +1,248 @@
+            app_version=__version__,
+        ),
+        house_key=entry.data.get(HOUSE_KEY),
+        token_listener=token_listener,
+    )
+
+    # Fetch devices
+    await manager.init_manager()
+    await manager.update_device_cache()
+    listener = DeviceListener(hass, manager)
+    manager.add_device_listener(listener)
+
+    ids = []
+    if hasattr(entry, "runtime_data") and entry.runtime_data is not None:
+        old_manager = entry.runtime_data.manager
+        ids = await compare_manager(old_manager, manager)
+
+    entry.runtime_data = HomeAssistantDuwiData(manager=manager, listener=listener)
+
+    # Clean up inappropriate devices
+    await cleanup_device_registry(hass, entry, ids)
+    hass.data[DOMAIN].setdefault("existing_house", []).append(entry.data[HOUSE_NO])
+
+    # Start global WebSocket listener
+    await hass.async_create_task(manager.ws.reconnect())
+
+    hass.loop.create_task(manager.ws.listen())
+    hass.loop.create_task(manager.ws.keep_alive())
+
+    await hass.config_entries.async_forward_entry_setups(entry, SUPPORTED_PLATFORMS)
+
+    return True
+
+
+async def compare_manager(old_manager: Manager, new_manager: Manager) -> list:
+    """Compare old and new manager to find devices with different room or type."""
+
+    ids = []
+
+    # Compare devices in both managers
+    for dev_id in old_manager.device_map:
+        d1 = old_manager.device_map.get(dev_id)
+        d2 = new_manager.device_map.get(dev_id)
+        if not d2:
+            # Device not found in new manager
+            continue
+
+        # Compare room and device type; if different, add device ID to the list
+        if d1.room_no != d2.room_no or d1.device_sub_type_no != d2.device_sub_type_no:
+            ids.append(dev_id)
+
+    return ids
+
+
+async def cleanup_device_registry(
+    hass: HomeAssistant, entry: DuwiConfigEntry, device_ids: list
+) -> None:
+    """Remove deleted device registry entry if there are no remaining entities."""
+
+    device_registry = dr.async_get(hass)
+
+    for dev_id, device_entry in list(device_registry.devices.items()):
+        for item in device_entry.identifiers:
+            if item[0] == DOMAIN and (
+                item[1] not in entry.runtime_data.manager.device_map
+                or dev_id in device_ids
+            ):
+                device_registry.async_remove_device(dev_id)
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: DuwiConfigEntry) -> bool:
+    """Unload the Duwi platforms associated with the provided config entry."""
+    if unload_ok := await hass.config_entries.async_unload_platforms(
+        entry, SUPPORTED_PLATFORMS
+    ):
+        house_no = entry.data.get(HOUSE_NO)
+        if house_no in hass.data[DOMAIN].get("existing_house"):
+            hass.data[DOMAIN].get("existing_house").remove(house_no)
+        await entry.runtime_data.manager.unload(True)
+
+    return unload_ok
+
+
+class DeviceListener(SharingDeviceListener):
```

Comment:
We use WebSocket to push updates to the device's status. For example, when a device's status changes, it will notify the `update_device` method of the `SharingDeviceListener`. In my implementation, the logic of the `update_device` method is to send a notification using `f"{DUWI_HA_SIGNAL_UPDATE_ENTITY}_{device.device_no}"`.

Each device inherits from `DuwiEntity`, and when the device entity is loaded, it listens for the update signal related to its specific device status:

```python
async_dispatcher_connect(
    self.hass,
    f"{DUWI_HA_SIGNAL_UPDATE_ENTITY}_{self.device.device_no}",
    self.handle_signal,
)
```

When the device status is updated, the Home Assistant device entity will reload the latest status of the corresponding device from the `manager` and then call `self.async_write_ha_state()` to update the device's state in Home Assistant.

Through this process, whenever the device's status changes, the Home Assistant entity's state will be updated accordingly.

Reply:
What's the payload that we get?

Reply:
Because each entity in Home Assistant maintains a reference to its associated device, when the WebSocket receives a device status update, it modifies the corresponding device in the manager. Since my device attribute is passed by reference, the device state in the Home Assistant entity is automatically updated.

Therefore, the manager only needs to call the update_device method of SharingDeviceListener to send an update signal. The corresponding device entity in Home Assistant will receive the signal and execute self.sync_write_ha_state() to refresh its state in Home Assistant.

I don't need to implement any additional logic to reload the device state, because the device in the entity is already up to date due to the reference type being used.

Reply:
![c455acbf025347dfaa53dfa31ba60d7](https://github.com/user-attachments/assets/f9b892f2-8164-4af3-9467-1a3ca12f58df)
![f31a0ae3851a65430019ffdf87a25f6](https://github.com/user-attachments/assets/1cc699d3-f76a-4ba4-9d42-1a9856e9b606)